runtime/src/accounts_db.rs

```diff
@@ -4252,69 +4252,110 @@ impl AccountsDb {
+                let loaded_write_version = loaded_account.write_version();
+                let loaded_hash = loaded_account.loaded_hash();
+                let should_insert = if let Some(existing_entry) = accum.get(loaded_account.pubkey())
+                {
+                    loaded_write_version > existing_entry.value().version()
+                } else {
+                    true
+                };
+                if should_insert {
+                    // Detected insertion is necessary, grabs the write lock to commit the write,
+                    match accum.entry(*loaded_account.pubkey()) {
+                        // Double check in case another thread interleaved a write between the read + write.
+                        Occupied(mut occupied_entry) => {
+                            if loaded_write_version > occupied_entry.get().version() {
+                                occupied_entry.insert((loaded_write_version, loaded_hash));
                             }
+                        }
 
-                            Vacant(vacant_entry) => {
-                                vacant_entry.insert((loaded_write_version, loaded_hash));
-                            }
+                        Vacant(vacant_entry) => {
+                            vacant_entry.insert((loaded_write_version, loaded_hash));
                         }
                     }
-                },
-            );
-        scan.stop();
+                }
+            },
+        )
+    }
 
-        let mut accumulate = Measure::start("accumulate");
-        let hashes: Vec<_> = match scan_result {
+    /// Given results of `scan_slot_for_dirty_pubkeys_and_hashes()`, map them to a Vec of Pubkey
+    /// and Hash
+    fn map_scan_result_to_pubkeys_and_hashes(
+        &self,
+        scan_result: ScanStorageResult<(Pubkey, Hash), DashMapVersionHash>,
+    ) -> Vec<(Pubkey, Hash)> {
+        match scan_result {
             ScanStorageResult::Cached(cached_result) => cached_result,
             ScanStorageResult::Stored(stored_result) => stored_result
                 .into_iter()
                 .map(|(pubkey, (_latest_write_version, hash))| (pubkey, hash))
-                .collect(),
-        };
-        let dirty_keys = hashes.iter().map(|(pubkey, _hash)| *pubkey).collect();
+                .collect::<Vec<_>>(),
+        }
+    }
 
+    /// Scan a slot in the account storage for dirty pubkeys and insert them into the list of
+    /// uncleaned pubkeys
+    ///
+    /// This function is called in Bank::drop() when the bank is _not_ frozen, so that its pubkeys
+    /// are considered for cleanup
+    pub fn scan_slot_and_insert_dirty_pubkeys_into_uncleaned_pubkeys(&self, slot: Slot) {
+        let scan_result = self.scan_slot_for_dirty_pubkeys_and_hashes(slot);
+        let mapped_result = self.map_scan_result_to_pubkeys_and_hashes(scan_result);
+        let dirty_pubkeys = mapped_result
+            .into_iter()
+            .map(|(pubkey, _hash)| pubkey)
```

Comment:
Hmmm instead of dumping the `hash` here after cloning it in the scan, could we pass a custom filter `map_account: impl Fn(LoadedAccount) -> T` into `scan_slot_for_dirty_pubkeys_and_hashes` to get the exact types we want

Reply:
Oh, that sounds cool! I don't actually know how to do that. Can you help me with a little more hand-holding?